File: python/day11.py
from aoc import *
from operator import mul, add, pow
from copy import deepcopy


class Monkey:
    lcm = 1

    def __init__(self, information):
        lines = information.splitlines()
        ops = lines[2].split()
        try:
            val = int(ops[-1])
            op = mul if ops[-2] == '*' else add
        except ValueError:
            val = 2
            op = pow
        d, t, f = mapl(first_int, lines[3:])
        Monkey.lcm *= d
        self.items = list(integers(lines[1]))
        self.operation = op
        self.value = val
        self.divisor = d
        self.true_case = t
        self.false_case = f
        self.inspected = 0

    def division_test(self, x):
        return self.false_case if x % self.divisor else self.true_case

    def inspect(self, worry_division):
        self.inspected += 1
        item = self.items.pop()
        worry = self.operation(item, self.value)
        if worry_division: worry //= 3
        worry %= Monkey.lcm
        return worry


def play_round(monkeys, worry_division):
    for monkey in monkeys:
        while monkey.items:
            worry_level = monkey.inspect(worry_division)
            where_to = monkey.division_test(worry_level)
            monkeys[where_to].items.append(worry_level)


def play_game(monkeys, rounds, worry_division):
    monkeys = deepcopy(monkeys)
    for _ in range(rounds):
        play_round(monkeys, worry_division)
    inspections = sorted(m.inspected for m in monkeys)
    return mul(*inspections[-2:])


def solve(file=11):
    monkeys = [Monkey(m) for m in read_input(file, sep="\n\n")]
    return play_game(monkeys, 20, True), play_game(monkeys, 10_000, False)


print(solve())










def test():
    assert solve("11_test") == (10605, 2713310158)
    assert solve() == (66802, 21800916620)

File: clojure/day11.clj
(ns day11
  (:require aoc
            [clojure.string :as str]
            [clojure.core.match :refer [match]]))


(defrecord Monkey [inspected items operation divisor test reducer])


(defn parse-operation [text]
  (match (vec text)
    ["*" "old"] #(* % %)
    ["+" "old"] #(+ % %)
    ["*" val]   #(* % (parse-long val))
    ["+" val]   #(+ % (parse-long val))))

(defn parse-monkey [lines]
  (let [items (aoc/integers (lines 1))
        op    (-> (lines 2)
                  (str/split #" ")
                  (#(take-last 2 %))
                  parse-operation)
        [divisor if-true if-false] (->> lines
                                        (drop 3)
                                        (map aoc/integers)
                                        (map first))]
    (map->Monkey
     {:inspected 0
      :items     items
      :operation op
      :divisor   divisor
      :test      (fn [x] (if (zero? ^long (mod x divisor)) if-true if-false))})))

(defn update-monkeys [m k1 k2 f]
  ;; Much faster than the `update-in` built-in.
  (let [m2 (m k1)]
    (assoc m k1 (assoc m2 k2 (f (k2 m2))))))

(defn monkey-play [monkeys curr]
  (loop [monkeys monkeys]
    (if-let [val (first (:items (monkeys curr)))]
      (let [{:keys [operation reducer test]} (monkeys curr)
            worry (reducer (operation val))
            dest  (test worry)]
        (recur (-> monkeys
                   (update-monkeys curr :inspected inc)
                   (update-monkeys curr :items rest)
                   (update-monkeys dest :items #(conj % worry)))))
      monkeys)))

(defn play-round [monkeys]
  (reduce monkey-play monkeys (range (count monkeys))))

(defn play-game [monkeys rounds divide-worry?]
  (let [lcm     (reduce * (map :divisor monkeys))
        reducer (if divide-worry? #(quot ^long % 3) #(mod ^long % lcm))]
    (->> monkeys
         (mapv #(assoc % :reducer reducer))
         (iterate play-round)
         (#(nth % rounds))
         (map :inspected)
         (sort >)
         (take 2)
         (reduce *))))

(defn parse-input [input]
  (->> input
       aoc/read-input-paragraphs
       (mapv parse-monkey)))

(defn solve
  ([] (solve 11))
  ([input]
   (let [monkeys (parse-input input)]
     [(play-game monkeys 20 true)
      (play-game monkeys 10000 false)])))


(solve)
